### DISCUSSION TITLE
NoClassDefFoundError when launching in game, works fine in IntelliJ

### DISCUSSION THREAD

**Discussion by Raynuz** [FabricMC/fabric-api#4020] [Mod Dev Support] [answered]
### Description

After fixing the problem with the mixin I had, I have now run into another problem: https://pastebin.com/D92Q4bTC

![image](https://github.com/user-attachments/assets/10016457-f4e7-4abd-9ede-a8d51a4a60ed)

When I try to launch it in game, it crashes with this error. However, it doesn't crash in the IDE I'm using (IntelliJ), and works perfectly. 

Would this be a problem with the way it's being built? For more debugging, please notify me of what files I should send. I'm not sure if this is the right place to ask as I'm new to all of this.

**Comment by Daomephsta** [comment, marked as answer]
It doesn't crash in your IDE because Gradle pulls in all the dependencies for you.
However those dependencies do not get put in the built JAR unless you specifically ask for that, because often that's not what you want.

There are two ways to package dependencies:

1. Fabric's [`include` Gradle configuration](https://fabricmc.net/wiki/documentation:fabric_loom#dependency_configurations) (a.k.a. jar-in-jar).
2. The [Gradle Shadow plugin](https://imperceptiblethoughts.com/shadow/introduction/)

`include` is the easiest to use, and the best option for packaging mods. It also works well for non-mod dependencies that have few dependencies of their own. The typical issues people have are:

* it only packages the specified dependency, not any transitive dependencies (dependencies of dependencies). This is because for its main usecase of packaging mods, transitivity would cause many accidental cases of packaging Minecraft and other provided libraries inside a mod JAR.
* it only causes the dependency to be packaged. It does not add it to the compile classpath or deobfuscate, as `modImplementation` does. It's meant to be used in addition, not as a replacement.

Gradle Shadow is harder to setup, requiring much more knowledge of Gradle and Loom. It's mainly good for packaging dependencies that aren't mods, and have many dependencies of their own.

**Reply by Raynuz** [reply]
Thank you so much, I've used the include method and it's working perfectly now without any crashes at all.
